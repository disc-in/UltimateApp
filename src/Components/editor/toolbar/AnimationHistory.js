import React, { useEffect, useState, useRef } from 'react';
import { StyleSheet, View, TouchableOpacity } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import theme from '../../../styles/theme.style';

// Stack of the values taken by a drill in the editor. Each time the drill is modified, a copy of the drill is added to the stack.
// Used to enable undo/redo actions
function AnimationHistory({ animation, onAnimationHistoryChange }) {
  const [currentId, setId] = useState(0);
  const [stack, setStack] = useState([animation]);

  const mustUpdate = useRef(false); // Flag used to trigger onAnimationHistoryChange with the correct animation

  useEffect(() => {
    if (!animation.isEqualTo(stack[currentId])) push();
  }, [animation]);

  useEffect(() => {
    if (mustUpdate.current) {
      mustUpdate.current = false;
      onAnimationHistoryChange(stack[currentId]);
    }
  }, [currentId]);

  // Add a drill to the stack (used when the user modifies the drill in the editor)
  const push = () => {
    // Remove all the drill status in the stack which are after the currently displayed status
    while (stack.length > currentId + 1) stack.pop();

    setStack([...stack, animation]);
    setId(currentId + 1);
  };

  // Restore the drill to its previous value
  const undo = () => {
    mustUpdate.current = true;
    setId(currentId - 1);
  };

  // Restore the drill to its next value
  const redo = () => {
    mustUpdate.current = true;
    setId(currentId + 1);
  };

  const canUndo = currentId !== 0;
  const canRedo = currentId !== stack.length - 1;
  return (
    <View style={styles.history}>
      <TouchableOpacity onPress={undo} disabled={!canUndo} style={styles.touchable} testID="undoButton">
        <MaterialCommunityIcons
          name="undo-variant"
          color={canUndo ? theme.COLOR_PRIMARY_LIGHT : theme.COLOR_SECONDARY}
          size={30}
        />
      </TouchableOpacity>
      <TouchableOpacity onPress={redo} disabled={!canRedo} style={styles.touchable} testID="redoButton">
        <MaterialCommunityIcons
          name="redo-variant"
          color={canRedo ? theme.COLOR_PRIMARY_LIGHT : theme.COLOR_SECONDARY}
          size={30}
        />
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  history: {
    flex: 1,
    flexDirection: 'row',
  },
  touchable: {
    flexGrow: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default AnimationHistory;
