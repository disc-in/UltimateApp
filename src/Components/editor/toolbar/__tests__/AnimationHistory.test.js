import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react-native';

import Drill from '../../../animation/Drill';

import AnimationHistory from '../AnimationHistory';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

describe('<AnimationHistory />', () => {
  afterEach(() => jest.clearAllMocks());

  const onAnimationHistoryChange = jest.fn();
  const animation1 = new Drill({ background: 'background1' });
  const animation2 = new Drill({ background: 'background2' });
  const animation3 = new Drill({ background: 'background3' });

  it('renders correctly', async () => {
    const { toJSON } = await waitFor(() =>
      render(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation1} />),
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('does undo and redo', async () => {
    const { rerender, getByTestId } = await waitFor(() =>
      render(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation1} />),
    );

    // Animation is changed in the editor
    await rerender(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation2} />);
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();

    await fireEvent.press(getByTestId('undoButton'));
    expect(onAnimationHistoryChange).toHaveBeenCalledWith(animation1);

    await fireEvent.press(getByTestId('redoButton'));
    expect(onAnimationHistoryChange).toHaveBeenCalledWith(animation2);
  });

  it('cannot undo if on the first animation', async () => {
    const { rerender, getByTestId } = await waitFor(() =>
      render(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation1} />),
    );

    // No changes in the editor
    await fireEvent.press(getByTestId('undoButton'));
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();

    // Animation is changed in the editor
    await rerender(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation2} />);
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();

    await fireEvent.press(getByTestId('undoButton'));
    expect(onAnimationHistoryChange).toHaveBeenCalledWith(animation1);

    // Already reached the beginning of history
    onAnimationHistoryChange.mockReset();
    await fireEvent.press(getByTestId('undoButton'));
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();
  });

  it('cannot redo if on the last animation', async () => {
    const { rerender, getByTestId } = await waitFor(() =>
      render(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation1} />),
    );

    // No changes in the editor
    await fireEvent.press(getByTestId('redoButton'));
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();

    // Animation is changed in the editor
    await rerender(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation2} />);
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();

    // Going to the beginning of history
    await fireEvent.press(getByTestId('undoButton'));
    onAnimationHistoryChange.mockReset();

    await fireEvent.press(getByTestId('redoButton'));
    expect(onAnimationHistoryChange).toHaveBeenCalledWith(animation2);

    // Already reached the end of history
    onAnimationHistoryChange.mockReset();
    await fireEvent.press(getByTestId('redoButton'));
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();
  });

  it('clears redo if animation changes while not at the end', async () => {
    const { rerender, getByTestId } = await waitFor(() =>
      render(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation1} />),
    );

    // Animation is changed in the editor
    await rerender(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation2} />);
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();

    await fireEvent.press(getByTestId('undoButton'));
    expect(onAnimationHistoryChange).toHaveBeenCalledWith(animation1);

    onAnimationHistoryChange.mockReset();
    // Animation is changed in the editor
    await rerender(<AnimationHistory onAnimationHistoryChange={onAnimationHistoryChange} animation={animation3} />);
    await fireEvent.press(getByTestId('redoButton'));
    expect(onAnimationHistoryChange).not.toHaveBeenCalled();
  });
});
