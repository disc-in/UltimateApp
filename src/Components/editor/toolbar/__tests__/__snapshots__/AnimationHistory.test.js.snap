// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimationHistory /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#A6A6A6",
            "fontSize": 30,
          },
          undefined,
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
      testID="undoButton"
    >
      󰕍
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#A6A6A6",
            "fontSize": 30,
          },
          undefined,
          Object {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
      testID="redoButton"
    >
      󰑏
    </Text>
  </View>
</View>
`;
